src/Credentials/EcsCredentialProvider.php

```diff
@@ -17,33 +20,41 @@ class EcsCredentialProvider
     const ENV_FULL_URI = "AWS_CONTAINER_CREDENTIALS_FULL_URI";
     const ENV_AUTH_TOKEN = "AWS_CONTAINER_AUTHORIZATION_TOKEN";
     const ENV_TIMEOUT = 'AWS_METADATA_SERVICE_TIMEOUT';
+    const ENV_RETRIES = 'AWS_METADATA_SERVICE_NUM_ATTEMPTS';
 
     /** @var callable */
     private $client;
 
+    /** @var int */
+    private $retries;
+
+    /** @var int */
+    private $attempts;
+
     /** @var float|mixed */
     private $timeout;
 
     /**
      *  The constructor accepts following options:
      *  - timeout: (optional) Connection timeout, in seconds, default 1.0
+     *  - retries: Optional number of retries to be attempted.
      *  - client: An EcsClient to make request from
      *
      * @param array $config Configuration options
      */
     public function __construct(array $config = [])
     {
-        $timeout = getenv(self::ENV_TIMEOUT);
-
-        if (!$timeout) {
-            $timeout = isset($_SERVER[self::ENV_TIMEOUT])
-                ? $_SERVER[self::ENV_TIMEOUT]
-                : (isset($config['timeout']) ? $config['timeout'] : 1.0);
-        }
-
-        $this->timeout = (float) $timeout;
+        $this->timeout = (float) getenv(self::ENV_TIMEOUT)
+            ?: (isset($config['timeout'])
+                ? $config['timeout']
+                : 1.0);
+        $this->retries = (int) getenv(self::ENV_RETRIES)
+            ?: (isset($config['retries'])
+                ? $config['retries']
+                : 3);
```

Comment:
Personally, I think it's more logical to to use a different fallback here:
1. Use `$config['timeout']`, otherwise
2. Use `getenv(self::ENV_TIMEOUT)`, otherwise
3. Fallback to `1.0`

Some environment variables may already be provided, outside of the application. So if you need multiple ECS credential providers, there is no way to set these configs per provider?

```suggestion
        $this->timeout = (int) isset($config['timeout'])
            ? $config['timeout']
            : (getenv(self::ENV_TIMEOUT) ?: 1.0);
        $this->retries = (int) isset($config['retries'])
            ? $config['retries']
            : (getenv(self::ENV_RETRIES) ?: 3);
```

What do you think?

Reply:
I agree that looking at the most explicit (config) first should be the way to go here.  Let's go with that and try to incorporate the null coalescing operator now that we've deprecated < 7.2